risc0/zkvm/src/receipt.rs

```diff
@@ -149,76 +157,58 @@ pub struct ReceiptMetadata {
-/// The [SessionReceipt] is implemented for receipts that represent
-/// [crate::Session]
-///
-/// There are several different flavors of receipts and this trait defines the
-/// the functions that all [crate::Session] receipts have such as accessing the
-/// journal and a way to verify the receipt.
-pub trait SessionReceipt: Debug {
-    /// Verifies the integrity of this receipt.
-    ///
-    /// Uses the ZKP system to cryptographically verify that each constituent
-    /// Segment has a valid receipt, and validates that these [SegmentReceipt]s
-    /// stitch together correctly, and that the initial memory image matches the
-    /// given `image_id` parameter.
-    #[cfg(not(target_os = "zkvm"))]
-    #[must_use]
-    fn verify(&self, image_id: Digest) -> Result<(), VerificationError>;
-
-    /// All session receipts have a journal
-    fn get_journal(&self) -> &Vec<u8>;
-
-    /// this provides a way to serialize a receipt
-    fn encode(&self) -> Vec<u8>;
-
-    /// get the length of the seal. This is used primarily for benchmarking
-    fn get_seal_len(&self) -> usize;
-
-    /// this is used for downcasting, primarily used for testing
-    #[cfg(test)]
-    fn as_any(&self) -> &dyn core::any::Any;
-}
-
-/// A free function that verifies the receipt
-///
-/// This function is a wrapper for [SessionReceipt::verify] that allows the
-/// caller to pass parameters that implement type conversion to [Digest] by
-/// calling the `into` function.
-#[cfg(not(target_os = "zkvm"))]
-pub fn verify(
-    receipt: &dyn SessionReceipt,
-    image_id: impl Into<Digest>,
-) -> Result<(), VerificationError> {
-    receipt.verify(image_id.into())
-}
-
 /// A receipt attesting to the execution of a Session.
 ///
-/// A SessionFlatReceipt attests that the `journal` was produced by executing a
+/// A SessionReceipt attests that the `journal` was produced by executing a
 /// [crate::Session] based on a specified memory image. This image is _not_
 /// included in the receipt and must be provided by the verifier when calling
-/// [SessionFlatReceipt::verify].
-#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
-pub struct SessionFlatReceipt {
-    /// The constituent [SegmentReceipt]s.
+/// [SessionReceipt::verify].
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub struct SessionReceipt {
+    /// The constituent [Receipt]s.
     ///
-    /// Together these can be used by [SessionFlatReceipt::verify] to
+    /// Together these can be used by [SessionReceipt::verify] to
     /// cryptographically prove that this full Session was faithfully executed.
-    pub segments: Vec<SegmentReceipt>,
+    pub segments: Vec<Box<dyn Receipt>>,
 
     /// The public data written by the guest in this Session.
     ///
     /// This data is cryptographically authenticated in
-    /// [SessionFlatReceipt::verify].
+    /// [SessionReceipt::verify].
     pub journal: Vec<u8>,
 }
 
+/// Provide common functionality implemented by all receipt types.
+#[dyn_partial_eq]
```

Comment:
I just wanted to point this out: using `dyn_partial_eq` implicitly enables downcasting. You can see it here: https://github.com/risc0/risc0/pull/666/files#diff-86c5437677f80d40396200163171134d15dc3e6eb257f74c37f382c39da52697R205

I initially added this in my draft PR so that we could get `partialeq` on the structs implementing this trait for the serialize/deserialize tests. I don't think it's a problem to expose this but I thought I'd point it out

Reply:
Ah ya I noticed that after I tried to add `as_any`, apparently that's already supported with `dyn_partial_eq`. Thanks for the heads up.